File: tests/r_parity/38_drdid.R
# StatsPAI doubly-robust DiD parity (R side) -- Module 38.
#
# Reads data/38_drdid.csv and runs DRDID::drdid_imp_panel with the
# same covariate. Tolerance: rel < 1e-3 on ATT; rel < 5e-2 on SE.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(DRDID)
})

MODULE <- "38_drdid"

df <- read_csv_strict(MODULE)

# DRDID::drdid_imp_panel expects: y1 = post outcome, y0 = pre outcome,
# D = treated indicator, covariates = pre-treatment covariate matrix.
# We have long-format data with one row per (id, post) -- pivot wide.
df_pre  <- df[df$post == 0, c("id", "y", "treated", "x")]
df_post <- df[df$post == 1, c("id", "y")]
names(df_pre)[names(df_pre) == "y"] <- "y0"
names(df_post)[names(df_post) == "y"] <- "y1"
wide <- merge(df_pre, df_post, by = "id")

fit <- DRDID::drdid_imp_panel(
  y1       = wide$y1,
  y0       = wide$y0,
  D        = wide$treated,
  covariates = cbind(1, wide$x)  # intercept + x
)

att <- fit$ATT
se  <- fit$se
ci_lo <- att - qnorm(0.975) * se
ci_hi <- att + qnorm(0.975) * se

rows <- list(
  parity_row(MODULE, "att",
             estimate = att, se = se, n = nrow(df)),
  parity_row(MODULE, "ci_lower",
             estimate = ci_lo, n = nrow(df)),
  parity_row(MODULE, "ci_upper",
             estimate = ci_hi, n = nrow(df))
)

write_results(MODULE, rows, extra = list(
  method = "DRDID::drdid_imp_panel",
  covariates = "x"
))

File: tests/r_parity/38_drdid.py
"""StatsPAI doubly-robust DiD parity (Python side) -- Module 38.

Generates a 2x2 DR-DID DGP with one covariate and runs sp.drdid
(improved DR estimator, Sant'Anna & Zhao 2020). The companion R
script uses DRDID::drdid_imp_panel.

Tolerance: rel < 1e-3 on ATT (closed-form once IF is fixed); rel <
5e-2 on the bootstrap/IF SE.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "38_drdid"


def make_data(n: int = 800, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    x = rng.normal(0, 1, n)
    # Logistic propensity in x
    ps = 1 / (1 + np.exp(-(0.3 + 0.5 * x)))
    G = rng.binomial(1, ps).astype(int)

    # 2x2 panel: one row per unit-period
    rows = []
    for i in range(n):
        Gi = int(G[i])
        xi = float(x[i])
        # Pre-period
        y_pre = 1.0 + 0.5 * xi + 0.2 * Gi + rng.normal(0, 1)
        rows.append({"id": i, "post": 0, "treated": Gi, "x": xi, "y": y_pre})
        # Post-period: true ATT = 2.0
        y_post = 1.0 + 0.5 * xi + 0.2 * Gi + 1.5 + 2.0 * Gi + rng.normal(0, 1)
        rows.append({"id": i, "post": 1, "treated": Gi, "x": xi, "y": y_post})
    return pd.DataFrame(rows)


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    res = sp.drdid(
        data=df, y="y", group="treated", time="post",
        covariates=["x"],
        method="imp",
        n_boot=200,
        random_state=PARITY_SEED,
    )

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "att",
                     estimate=float(res.estimate),
                     se=float(res.se),
                     n=int(len(df))),
    ]
    # CI bounds
    rows.append(ParityRecord(
        MODULE, "py", "ci_lower",
        estimate=float(res.ci[0]), n=int(len(df))))
    rows.append(ParityRecord(
        MODULE, "py", "ci_upper",
        estimate=float(res.ci[1]), n=int(len(df))))

    write_results(
        MODULE, "py", rows,
        extra={
            "method": "DR-DID (improved, Sant'Anna-Zhao 2020)",
            "covariates": ["x"],
            "n_boot": 200,
        },
    )


if __name__ == "__main__":
    main()
